packages/hash/frontend/src/blocks/page/BlockContextMenu/BlockContextMenuItem.tsx

```diff
@@ -0,0 +1,93 @@
+import { faChevronRight } from "@fortawesome/free-solid-svg-icons";
+import { ListItemIcon, ListItemText, MenuItem } from "@mui/material";
+import {
+  bindPopover,
+  bindHover,
+  bindFocus,
+  usePopupState,
+} from "material-ui-popup-state/hooks";
+import HoverPopover from "material-ui-popup-state/HoverPopover";
+import { cloneElement, forwardRef, RefObject, useRef, useState } from "react";
+import { FontAwesomeIcon } from "../../../shared/icons";
+
+type BlockContextMenuItemProps = {
+  itemKey: string;
+  onClick?: () => void;
+  icon: JSX.Element;
+  title: JSX.Element | string;
+  subMenu?: JSX.Element;
+  subMenuWidth?: number;
+};
+
+export const BlockContextMenuItem = forwardRef<
+  HTMLLIElement,
+  BlockContextMenuItemProps
+>(({ onClick, icon, title, itemKey, subMenu, subMenuWidth }, ref) => {
+  const subMenuPopupState = usePopupState({
+    variant: "popper",
+    popupId: `${itemKey}-submenu`,
+  });
+  const [subMenuOffsetTop, setSubmenuOffsetTop] = useState<
+    number | undefined
+  >();
+  const localRef = useRef<HTMLLIElement>(null);
+  const menuItemRef = (ref ?? localRef) as RefObject<HTMLLIElement>;
+
+  if (subMenu && !subMenuOffsetTop && menuItemRef.current) {
+    setSubmenuOffsetTop(menuItemRef.current.offsetTop);
+  }
```

Comment:
In the designs, submenus are aligned to the top

![Screenshot 2022-04-14 at 01 19 47](https://user-images.githubusercontent.com/16928688/163290174-b9edd19b-4865-4a8e-a84a-1bc2a93d1c2f.png)

I need that code to get the necessary offset needed to achieve the design. Without it, it looks this way

![Screenshot 2022-04-14 at 01 26 49](https://user-images.githubusercontent.com/16928688/163290802-5bf496ad-ce47-4ee5-9175-184e8f92f2f8.png)


I want it to be triggered the first time `BlockContextMenuItem` is rendered (which is when we open BlockContextMenu). It's only triggered for MenuItems that have submenu's. Placed it in the render body instead of an effect to save a re-render using the technique you once talked about. Let me know if there is a better way this can be implemented.

Reply:
It's definitely improper for this to be in the render body. The technique I mentioned in the past is to do with deriving state from props, not to do with interacting with the DOM. I would implement this using `useLayoutEffect`.

Reply:
I see. Thanks for explaining